lib/common/recurrence.dart

```diff
@@ -1,35 +1,98 @@
 class Recurrence {
-  bool enabled;
-  DateTime? timeStart;
-  DateTime? timeEnd;
+  late bool enabled;
+  late DateTime timeStart;
+  late DateTime timeEnd;
   // monday, tuesday, ... , saturday, sunday
-  List<bool>? dates;
-  int? id; // ideally this should be final
+  late List<bool> dates;
+  late String id; // ideally this should be final
 
   Recurrence(
+      {this.enabled = false,
+        DateTime? timeStart,
+        DateTime? timeEnd,
+        List<bool>? dates,
+        String? id}) {
+    this.timeStart = timeStart ?? DateTime(2000);
+    this.timeEnd = timeEnd ?? DateTime(2000);
+    this.dates = dates ?? [false, false, false, false, false, false, false];
+    this.id = id ?? DateTime.now().millisecondsSinceEpoch.toString();
+  }
+  
+  /// Set nullOrId: null to set id to auto-generate an id instead
+  Recurrence.requireFields(
       {required this.enabled,
         required this.timeStart,
         required this.timeEnd,
         required this.dates,
-        this.id}) {
-    id = id ?? DateTime.now().millisecondsSinceEpoch;
+        required String? nullOrId}) {
+    this.id = nullOrId ?? DateTime.now().millisecondsSinceEpoch.toString();
   }
 
-  Recurrence.requireFields(
-      {required this.enabled,
-        required this.timeStart,
-        required this.timeEnd,
-        required this.dates}) {
-    id ??= DateTime.now().millisecondsSinceEpoch; // something about this design pattern looks off... not sure how to make this better since flutter won't allow me to put an optional named argument in
+  Recurrence.fromMap(Map<String, dynamic>? recurrenceRulesMap, {String? extraId}) {
+    if (recurrenceRulesMap == null) {
+      enabled = false;
+      timeStart = DateTime(2000);
+      timeEnd = DateTime(2000);
+      dates = [false, false, false, false, false, false, false];
+      id = DateTime.now().millisecondsSinceEpoch.toString();
+    } else {
+      enabled = recurrenceRulesMap?['enabled'] ?? false;
+      timeStart = recurrenceRulesMap?["starts on"].toDate() ?? recurrenceRulesMap?["starts on"] ?? DateTime(2000);
+      timeEnd = recurrenceRulesMap?["ends on"].toDate() ?? recurrenceRulesMap?["ends on"] ?? DateTime(2000);
+      dates = [];
+      // recurrenceRulesMap!['dates'].forEach((tag) {dates.add(tag as bool);}); // List<dynamic> bullshit
+      if (recurrenceRulesMap?["repeat on days"]!=null) { 
+        for (var b in recurrenceRulesMap!["repeat on days"]) {
+          dates.add(b as bool);
+      }
+      } else {
+          dates = [false, false, false, false, false, false, false];
+      }
+      id = extraId ?? (recurrenceRulesMap?.containsKey("id")!=null ? recurrenceRulesMap!["id"] : DateTime.now().millisecondsSinceEpoch.toString());
+    }
   }
+
   toMap() {
     return ({
       'enabled' : enabled,
-      'starts on' : timeStart,
+      'starts on' : timeStart, // not in format
       'ends on' : timeEnd,
       'repeat on days' : dates,
-      'id' : id
+      'id' : id // not in format
```

Comment:
not in our example format but it should be automatically added when written

Reply:
Yeah, its in the example format now too. Ill remove this comment